worker/upload.go

```diff
@@ -101,21 +78,25 @@ type (
 		mem              *acquiredMemory
 		overdriveTimeout time.Duration
 
-		candidates []*uploader // sorted by upload estimate
+		sectors    []*sectorUpload
+		candidates []*candidate // sorted by upload estimate
 		shards     [][]byte
 
-		mu          sync.Mutex
-		numInflight uint64
-		numLaunched uint64
-		numUploaded uint64
-
-		overdriving    map[int]map[types.PublicKey]struct{}
-		lastOverdrive  time.Time
+		mu             sync.Mutex
+		numInflight    uint64
+		numLaunched    uint64
 		numOverdriving uint64
+		numUploaded    uint64
+
+		lastOverdrive time.Time
+		errs          HostErrorSet
+	}
+
+	candidate struct {
+		uploader *uploader
 
-		sectors map[int]*sectorUpload
-		used    map[types.PublicKey]struct{}
-		errs    HostErrorSet
+		used        bool
+		overdriving int // sector index
```

Comment:
I believe the only reason `sectorUpload.index` exists at this point is to compare it to this `overdriving` field.
Why not merge `used` and `overdriving` into a `*sectorUploadRequest` field.
If it's `nil`, the candidate is not used and if it is not `nil` it is used to handle the request.
Then you can use the `overdrive` field on the request to determine whether the candidate is an overdrive candidate or not.